### PR TITLE
MM-47228 - restrict guest invitation flow if subscription plan does n…

### PR DESCRIPTION
**Pull request by pvev**
#### Summary
Restrict guest invitation flow if the subscription plan does not support it or is not the free trial

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-47228

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: api4/team.go [hunk 1]
```diff
@@ -1485,6 +1485,29 @@ func inviteGuestsToChannels(c *Context, w http.ResponseWriter, r *http.Request)
...
+			return
+		}
+
+		isPaidSubscription := product.SKU == model.CloudSubscriptionSkuEnterprise || product.SKU == model.CloudSubscriptionSkuProfessional
+
+		if subscription.IsFreeTrial == "false" && !isPaidSubscription {
+			c.Err = model.NewAppError("Api4.InviteGuestsToChannels", "api.team.invite_guests_to_channels.disabled.error", nil, "", http.StatusForbidden)
+			return
+		}
+	}
```
Comment (marianunez) [inline api4/team.go:1509]: @pvev we should do a general check on license given that the guest restriction should be for both cloud and self-managed - if you are not licensed then you shouldn't be able to allow to invite guests

**Review by marianunez — CHANGES_REQUESTED** [review]

File: api4/team.go (hunk 1 above)
Comment (marianunez) [reply api4/team.go:1509]: It seems we are doing the check in another endpoint to invite guest directly here: https://github.com/mattermost/mattermost-server/blob/master/api4/team.go#L1464

File: api4/team.go [hunk 2]
```diff
@@ -1485,6 +1485,29 @@ func inviteGuestsToChannels(c *Context, w http.ResponseWriter, r *http.Request)
 		return
 	}
 
+	// restrict guest invites depending on the subscription type (see MM-47228)
+	isCloud := c.App.Srv().License() != nil && *c.App.Srv().License().Features.Cloud
```
Comment (neallred) [inline api4/team.go:1489]: I think we should instead check for the feature, e.g. `if c.App.Channels().License() != nil && *c.App.Channels().License().Features.GuestAccounts` . That way, both cloud and self hosted are covered. I think it was just a miss that we never added that check. Then I think we can omit all the subscription and product checking.

**Review by neallred — CHANGES_REQUESTED** [review]

File: api4/team.go [hunk 3]
```diff
@@ -1462,12 +1462,12 @@ func inviteUsersToTeam(c *Context, w http.ResponseWriter, r *http.Request) {
...
 	graceful := r.URL.Query().Get("graceful") != ""
 	if c.App.Channels().License() == nil {
-		c.Err = model.NewAppError("Api4.InviteGuestsToChannels", "api.team.invate_guests_to_channels.license.error", nil, "", http.StatusNotImplemented)
+		c.Err = model.NewAppError("Api4.InviteGuestsToChannels", "api.team.invite_guests_to_channels.license.error", nil, "", http.StatusNotImplemented)
 		return
 	}
 
 	if !*c.App.Config().GuestAccountsSettings.Enable {
-		c.Err = model.NewAppError("Api4.InviteGuestsToChannels", "api.team.invate_guests_to_channels.disabled.error", nil, "", http.StatusNotImplemented)
+		c.Err = model.NewAppError("Api4.InviteGuestsToChannels", "api.team.invite_guests_to_channels.disabled.error", nil, "", http.StatusNotImplemented)
```
Comment (neallred) [inline api4/team.go:1470]: :+1:

File: model/license.go [hunk 4]
```diff
@@ -326,6 +326,25 @@ func NewTestLicense(features ...string) *License {
 	return ret
 }
 
+// NewTestLicense returns a license that expires in the future and set as false the given features.
+func NewTestLicenseWithFalseDefaults(features ...string) *License {
```
Comment (neallred) [inline model/license.go:330]: :+1:

**Review by neallred — APPROVED** [review]
Nice add on the unit test helper. LGTM now. Thanks, @pvev!

**Review by marianunez — APPROVED** [review]
Thanks @pvev!

**Review by srkgupta — CHANGES_REQUESTED** [review]
Thanks @pvev. Requested some additional changes. Please see the channel for more information.

**Review by srkgupta — APPROVED** [review]
Thanks @pvev. The changes LGTM.

**Pull request merged**